dracut/99setup-root/initrd-setup-root

```diff
@@ -7,10 +8,52 @@
 # files if needed and make it easier to debug. In fact, it was required
 # to use first_boot=1 with PXE to get a proper rootfs which is strange.
 
+function usrbin() {
+  LD_LIBRARY_PATH=/sysusr/usr/lib64 /sysusr/usr/bin/"$@"
+}
+
+function walksysroot() {
+  local topdir="$1"
+  shift
+  local action="$1"
+  shift
+  while IFS= read -r -d '' entry; do
+    "${action}" "${entry}" || true
+  done < <(chroot /sysroot /usr/bin/find "${topdir}" -xdev "$@" -print0 | usrbin tac -s "")
+  # Do a chroot to be able to pass "-regex /etc/something" without having to prepend /sysroot
+  # in the regex. Also do the print0 as last action, after filtering.
+  # Use tac to process deeper files/dirs before upper dirs.
+}
+
 # /etc/machine-id after a new image is created:
 COREOS_BLANK_MACHINE_ID="42000000000000000000000000000042"
 MACHINE_ID_FILE="/sysroot/etc/machine-id"
 
+function selectiveosreset() {
+  local entry="/sysroot$1"
+  # Don't remove /sysroot itself
+  [ "${entry}" = "/sysroot" ] && return
+  [ "${entry}" = "/sysroot/" ] && return
+  # Don't remove the active /usr mount point
+  [ "${entry}" = "/sysroot/usr" ] && return
+  if [ -d "${entry}" ]; then
+    # Try to delete dir, will fail if its contents are preserved
+    usrbin rmdir "${entry}" 2>/dev/null || true
+  else
+    rm "${entry}" || true # TODO: Report through flag file or fail the boot?
+  fi
+}
+
+# Do the selective OS reset as prepared by flatcar-reset
+if [ "$(cat /sysroot/selective-os-reset 2>/dev/null || true)" != "" ]; then
+  walksysroot / selectiveosreset -regextype egrep -not -regex "$(cat /sysroot/selective-os-reset)" || true
```

Comment:
I'm probably missing something but will this remove files from /etc if needed? We are passing -xdev to find and /etc is going to be a mountpoint, so this means that /etc will be untouched (whether we want or not).

Reply:
`/etc` is mounted at the end, first comes the OS reset, then the overlay cleanup and both operate on the folder `/etc` (the upperdir).